csharp/src/Ice/Reference.cs

```diff
@@ -532,27 +441,21 @@ public virtual bool Equals(Reference? other)
                                                 nameof(invocationTimeout));
                 }
 
-                reference = Clone(reference);
+                reference = CloneIfSame(reference);
                 reference._invocationTimeout = invocationTimeoutValue;
             }
             return reference;
         }
 
-        public Reference Clone()
-        {
-            var clone = (Reference)MemberwiseClone();
-            clone._requestHandler = null;
-            clone._requestHandlerMutex = null;
-            return clone;
-        }
+        internal virtual Reference Clone() => (MemberwiseClone() as Reference)!;
 
         // Creates a new clone of this unless `clone` is already a real clone (meaning "not this")
         // Returns the clone.
-        private protected T Clone<T>(T clone) where T : Reference
+        private protected T CloneIfSame<T>(T clone) where T : Reference
         {
             if (ReferenceEquals(this, clone))
             {
-                clone = (T)Clone();
+                clone = (Clone() as T)!;
```

Comment:
What is our guideline here? Do we remove explicit casting in favor of these 2 operators? I was going to change a cast like in the outgoing connection factory to also use "as" but after reading few recommandations I kept the explicit cast (operator as is better suited if you're not sure the object is of the specified type).

Reply:
I think `as` is better than cast unless you want user-defined casts.